From 2bd330b1375b53080ca50d7a6038d704527a683e Mon Sep 17 00:00:00 2001
From: Christophe Roullier <christophe.roullier@st.com>
Date: Fri, 2 Dec 2022 10:58:52 +0100
Subject: [PATCH 0136/1141] net: phy: realtek: add property to disable EEE for
 RTL8211F

Add property "realtek,eee-disable" to have possibility to disable
EEE (Energy-Efficient Ethernet) feature in Realtek Phy.
This feature is activated by default and in some usecases (TSN, gPTP etc..)
gPTP peer delays of 18us is observed when EEE is activated.
IEEE802.1AS mention to consider a system as AScapable; the peer delay
should be around 800ns. So in this situation we are hugely out of spec
and so no gPTP compliant. Without EEE, delay is improve.

Change-Id: I587aa104f8e4d45f87b18e870cdfbf222b6c9058
Signed-off-by: Christophe Roullier <christophe.roullier@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/278943
Tested-by: Eric FOURMONT <eric.fourmont-ext@st.com>
Domain-Review: Eric FOURMONT <eric.fourmont-ext@st.com>
Reviewed-by: Eric FOURMONT <eric.fourmont-ext@st.com>
---
 drivers/net/phy/realtek.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index d4b135c675bd..7832b1b4ad55 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -411,6 +411,16 @@ static int rtl8211f_config_init(struct phy_device *phydev)
 			val_rxdly ? "enabled" : "disabled");
 	}
 
+	if (of_property_read_bool(dev->of_node, "realtek,eee-disable")) {
+		rtl821x_write_page(phydev, 0xa4b);
+		phy_write(phydev, 0x11, 0x1110);
+		rtl821x_write_page(phydev, 0);
+		phy_write(phydev, 0xd,7);
+		phy_write(phydev, 0xe,0x3c);
+		phy_write(phydev, 0xd,0x4007);
+		phy_write(phydev, 0xe,0x0);
+	}
+
 	if (priv->has_phycr2) {
 		ret = phy_modify_paged(phydev, 0xa43, RTL8211F_PHYCR2,
 				       RTL8211F_CLKOUT_EN, priv->phycr2);
-- 
2.39.2

